tiledb/sm/c_api/tiledb_helpers.h

```diff
@@ -0,0 +1,356 @@
+    tiledb_ctx_t* ctx,
+    const tiledb_array_schema_evolution_t* schema_evolution) {
+  if (schema_evolution == nullptr ||
+      schema_evolution->array_schema_evolution_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB array schema evolution object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+inline int32_t sanity_check(tiledb_ctx_t* ctx, const tiledb_domain_t* domain) {
+  if (domain == nullptr || domain->domain_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB domain object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+inline int32_t sanity_check(tiledb_ctx_t* ctx, const tiledb_query_t* query) {
+  if (query == nullptr || query->query_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB query object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+inline int32_t sanity_check(
+    tiledb_ctx_t* ctx, const tiledb_query_condition_t* cond) {
+  if (cond == nullptr || cond->query_condition_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB query condition object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+inline int32_t sanity_check(tiledb_ctx_t* ctx, const tiledb_vfs_t* vfs) {
+  if (vfs == nullptr || vfs->vfs_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB virtual filesystem object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+inline int32_t sanity_check(tiledb_ctx_t* ctx, const tiledb_vfs_fh_t* fh) {
+  if (fh == nullptr || fh->vfs_fh_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB virtual filesystem file handle");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+inline int32_t sanity_check(
+    tiledb_ctx_t* ctx, const tiledb_fragment_info_t* fragment_info) {
+  if (fragment_info == nullptr || fragment_info->fragment_info_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB fragment info object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+inline int32_t sanity_check(tiledb_ctx_t* ctx, const tiledb_file_t* file) {
+  if (file == nullptr || file->file_ == nullptr) {
+    auto st = Status_Error("Invalid TileDB file object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+#endif
+
+inline int32_t check_filter_type(
+    tiledb_ctx_t* ctx, tiledb_filter_t* filter, tiledb_filter_type_t type) {
+  auto cpp_type = static_cast<tiledb::sm::FilterType>(type);
+  if (filter->filter_->type() != cpp_type) {
+    auto st = Status_FilterError(
+        "Invalid filter type (expected " + filter_type_str(cpp_type) + ")");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  }
+  return TILEDB_OK;
+}
+
+#if 01
```

Comment:
Remove. Code here needs to use the `api_entry` exception safety mechanism.

Reply:
preprocessor kruft removed
re api_entry, the routines exposed externally are using that mechanism, 
not sure what you're referring to otherwise, the check_() routines here were relocated from tiledb.cc to be available to more than one file... ?